sql: fix TRUNCATE for temporary tables

otan opened the pull request:
Resolves https://github.com/cockroachdb/cockroach/issues/47030

Release note (bug fix): TRUNCATE can now run on temporary tables,
fixing a bug in 20.1.0 where temporary tables could not be truncated,
resulting in an error `unexpected value: nil`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47482)
<!-- Reviewable:end -->

rohany commented on pkg/sql/logictest/testdata/logic_test/temp_table:
```diff
@@ -222,3 +222,21 @@ subtest table_with_on_commit
+
+subtest regression_47030
+
+statement ok`
```
got a hanging quote

otan replied:
i can't believe i've done this

rohany reviewed (approved):
I feel like we need to remove `MakePublicTableKey` -- all of its uses are potential breaking points for temp tables.

otan committed:
sql: fix TRUNCATE for temporary tables

Release note (bug fix): TRUNCATE can now run on temporary tables,
fixing a bug in 20.1.0 where temporary tables could not be truncated,
resulting in an error `unexpected value: nil`.

otan commented:
bors r=rohany
